### ISSUE TITLE
Auto-merge dependency updates

### ISSUE THREAD

**Issue by rarkins** [singapore/renovate#109]
There are some dependencies - especially in devDependencies - where I never do any further manual testing if they pass automated testing. In such a case, it would be nice to have them auto-merged.

This still needs requirements definition though, as it may not be so simple. e.g. GitHub has concepts like requiring a "review" before merge, etc.

**Comment by rarkins** [comment]
Waiting on #133

**Comment by rarkins** [comment]
* if no branch exists then create branch first (no PR)
* if no auto merge then raise PR
* if branch exists then check status 
* if status approved then merge
* if status failed then PR?

**Comment by rarkins** [comment]
I think that there may be multiple approaches people want to take to "auto-merging" so this will likely be an enumerated configuration type and not simply true/false. Default behaviour would remain as "none".

In the below I'm going to focus on GitHub.

The most obvious auto-merge workflow would be like this:

1. Check if branch exists. If not, then create it and exit.
2. If branch exists, [check its combined status](https://developer.github.com/v3/repos/statuses/#get-the-combined-status-for-a-specific-ref)
3. If branch combined status is "success", then merge it

This leaves questions:
- What to do if the branch status is "failure"? 
   * This will probably need to be configurable too. 
   * e.g. create PR or issue or do nothing
- How to merge?
   * Without a PR, there's no way to do this simply through the GitHub API
   * If you create a PR, merge, then delete it, the user will still get GitHub notifications (undesirable but not terrible)
   * If the renovate branch is up to date with master, we could simply rewrite the base ref (e.g. `master`) via the low-level git API
   * We could potentially re-apply the commit - to `master` - even if the renovate branch is not up-to-date, but I think this is risky. 

Note 1: Some CI systems can test branches against `master` too however I think this is only done once a PR is raised. Also, it is not continuously re-tested every time the base branch is updated with new commits

Note 2: The most careful way to do this would be to make sure that branches are always up-to-date with master before auto-merging them. e.g. to make sure no conflicts with other "simultaneous" dependency upgrades. This would 

Note 3: Need to consider various GitHub master branch PR protection mechanisms:
 - PR must be raised
   * Cannot commit directly to `master`, must commit via PR merging
 - Reviews must be added
   * Requires manual intervention or admin override, or another bot account to auto-approve
 - Branch must be up-to-date with master
   * Requires rebasing every time `master` changes (today we only do it if a merge conflict arises)

**Comment by rarkins** [comment]
My own (initial) requirement for some personal projects would be to auto-merge selected project/dependencies without PR and after all tests pass. This would mean such projects could not have the GitHub requirement to have reviews. To be safe, I'd want the branch to be up-to-date with master before merging, which could mean a lot of rebasing if multiple dependencies are updated at similar times. However, this auto-merge feature itself would reduce the number of times that dependency updates overlap in time.

**Comment by rarkins** [comment]
@hbetts it sounds like you would get a *lot* of `renovate` dependency updates each week across all your projects. Do you have any ideas/comments?

Considering GitLab:
- Rebasing branches via API is not possible so branches would need to be deleted and then recreated if you desire all dependency updates to be tested against latest `master`. 
- It's not possible via API to do a simple `master` git ref update to point to a tested `renovate` branch git ref, so therefore any automatic merges (without PR) would need to be done via a manual re-application of the change to `master`. I don't see this as being much of a risk though.

**Comment by rarkins** [comment]
@guumaster please let me know if this feature would be nice to have for you too and if you have any ideas/comments on the above.

**Comment by rarkins** [comment]
I suspect people may like granular control over which type to upgrade, e.g. a setting `automerge` with default value `none` and other values of `patch`, `minor` and `major`. This could for example allow people to apply a repository rule of `minor` but override it for certain dependencies to `patch` if those dependencies are known to make backwards-incompatible changes in minor version updates.

**Comment by guumaster** [comment]
What you mention, together with the grouping feature, will ease repetition a lot. Instead of going through lots of open PRs and manually merge them, you just would have to check that everything is merged, tested and running.

It'd be a very nice feature indeed

**Comment by rarkins** [comment]
In the meantime I've added support for a new configuration option [prCreation](https://github.com/singapore/renovate/commit/ca7b25825df7f0d0e2a69f9e750771c6574d6542)

It lets you configure whether to create PRs:
- immediately (current behaviour)
- only when status checks pass (so will suppress PRs for any breaking updates, but keep the branch)
- only once status checks are completed (status !== pending)

I think I will probably use `'not-pending'` for most of my projects, so that I'm not alerted to PRs before they've finished their tests.

**Comment by rarkins** [comment]
- [ ] [Autodetect GitHub repo supported merge types](https://developer.github.com/changes/2016-09-26-pull-request-merge-api-update/)
- [ ] Support GitLab
- [ ] Plan for future branch merging vs PR merging
